File: samples/dotnet-aspnet/src/ReportRepository.cs
using Microsoft.Data.SqlClient;

namespace Reporting;

/// <summary>
/// Report lookup over ADO.NET.
///
/// Intentionally vulnerable, see cases/sql-injection-concat.
/// </summary>
public class ReportRepository
{
    private readonly CryptoService _crypto;

    public ReportRepository(CryptoService crypto) => _crypto = crypto;

    /// <summary>
    /// Return every report belonging to an owner.
    ///
    /// The owner name arrives from a request parameter and is concatenated into the statement.
    /// </summary>
    public async Task<List<string>> FindByOwnerAsync(string owner)
    {
        var titles = new List<string>();

        await using var connection = new SqlConnection(_crypto.GetConnectionString());
        await connection.OpenAsync();

        // VULN: a quote in owner ends the literal and the remainder executes as SQL.
        string sql = "SELECT title FROM reports WHERE owner = '" + owner + "'";
        await using var command = new SqlCommand(sql, connection);

        await using var reader = await command.ExecuteReaderAsync();
        while (await reader.ReadAsync())
        {
            titles.Add(reader.GetString(0));
        }

        return titles;
    }
}

File: samples/java-spring/src/main/java/com/example/reporting/ReportRepository.java
package com.example.reporting;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;
import javax.sql.DataSource;
import org.springframework.stereotype.Repository;

/**
 * Report lookup over plain JDBC.
 *
 * Intentionally vulnerable, see cases/sql-injection-concat.
 */
@Repository
public class ReportRepository {

    private final DataSource dataSource;

    public ReportRepository(DataSource dataSource) {
        this.dataSource = dataSource;
    }

    /**
     * Return every report belonging to an owner.
     *
     * The owner name arrives from a request parameter and is concatenated into the statement.
     */
    public List<String> findByOwner(String owner) throws Exception {
        List<String> titles = new ArrayList<>();

        try (Connection connection = dataSource.getConnection();
                Statement statement = connection.createStatement()) {

            // VULN: a quote in owner ends the literal and the remainder is executed as SQL.
            String sql = "SELECT title FROM reports WHERE owner = '" + owner + "'";
            ResultSet results = statement.executeQuery(sql);

            while (results.next()) {
                titles.add(results.getString("title"));
            }
        }

        return titles;
    }
}
